File: ios/TiHTTPStreamProxy.swift
//
//  TiHTTPStreamProxy.swift
//  TiNetworkManager
//
//  Created by Douglas Alves on 02/01/26.
//


/**
 * Ti.Network.Manager - Stream Proxy
 * Proxy for streaming HTTP requests
 * Exposed to JavaScript for creating streaming connections (SSE, AI APIs, etc.)
 */

import TitaniumKit
import Foundation

@objc(TiHTTPStreamProxy)
class TiHTTPStreamProxy: TiProxy {
    
    // MARK: - Properties
    
    private var streamManager: TNMStreamManager
    private var interceptorManager: TNMInterceptorManager
    private var certificatePinningManager: TNMCertificatePinningManager
    
    private var requestId: String
    private var url: String
    private var method: String
    private var headers: [String: String]
    private var body: String?
    private var priority: String
    
    private var isActive = false
    private var startTime: Date?
    
    // MARK: - Initialization
    
    init(
        params: [String: Any],
        streamManager: TNMStreamManager,
        interceptorManager: TNMInterceptorManager,
        certificatePinningManager: TNMCertificatePinningManager
    ) {
        self.streamManager = streamManager
        self.interceptorManager = interceptorManager
        self.certificatePinningManager = certificatePinningManager
        
        // Generate unique request ID
        self.requestId = UUID().uuidString
        
        // Extract parameters with proper unwrapping
        self.url = params["url"] as? String ?? ""
        self.method = (params["method"] as? String ?? "GET").uppercased()
        self.headers = params["headers"] as? [String: String] ?? [:]
        self.body = params["body"] as? String
        self.priority = params["priority"] as? String ?? "normal"
        
        super.init()
        
        TNMLogger.debug("Stream proxy created", feature: "Streaming", details: [
            "requestId": requestId,
            "url": url,
            "method": method,
            "priority": priority
        ])
    }
    
    // MARK: - Public API
    
    /**
     * Start the streaming request
     * 
     * @param arguments Unused arguments array (required by Titanium)
     */
    @objc(start:)
    func start(arguments: [Any]?) {
        guard !isActive else {
            TNMLogger.warning("Stream already started", feature: "Streaming", details: [
                "requestId": requestId
            ])
            return
        }
        
        guard let url = URL(string: self.url) else {
            TNMLogger.error("Invalid URL for streaming request", feature: "Streaming", details: [
                "requestId": requestId,
                "url": self.url
            ])
            fireEvent("error", with: ["error": "Invalid URL"])
            return
        }
        
        isActive = true
        startTime = Date()
        
        TNMLogger.Streaming.started(requestId: requestId, url: self.url)
        
        // Apply request interceptors
        var modifiedHeaders = headers
        var modifiedBody = body
        
        interceptorManager.interceptRequest(
            url: self.url,
            method: method,
            headers: &modifiedHeaders,
            body: &modifiedBody
        )
        
        // Convert priority
        let urlPriority: Float
        switch priority {
        case "high":
            urlPriority = URLSessionTask.highPriority
        case "low":
            urlPriority = URLSessionTask.lowPriority
        default:
            urlPriority = URLSessionTask.defaultPriority
        }
        
        TNMLogger.debug("Stream priority set", feature: "Streaming", details: [
            "requestId": requestId,
            "priority": priority
        ])
        
        // Convert body to Data
        var bodyData: Data?
        if let bodyString = modifiedBody {
            bodyData = bodyString.data(using: .utf8)
        }
        
        // Get certificate validator if needed
        let certificateValidator = certificatePinningManager.getValidator(for: url.host ?? "")
        
        // Start streaming
        streamManager.startStream(
            requestId: requestId,
            url: url,
            method: method,
            headers: modifiedHeaders,
            body: bodyData,
            priority: urlPriority,
            certificateValidator: certificateValidator,
            onChunk: { [weak self] chunk in
                self?.handleChunk(chunk)
            },
            onComplete: { [weak self] statusCode, headers in
                self?.handleComplete(statusCode: statusCode, headers: headers)
            },
            onError: { [weak self] error in
                self?.handleError(error)
            }
        )
    }
    
    /**
     * Cancel the streaming request
     * 
     * @param arguments Unused arguments array (required by Titanium)
     */
    @objc(cancel:)
    func cancel(arguments: [Any]?) {
        guard isActive else { return }
        
        TNMLogger.Streaming.cancelled(requestId: requestId)
        
        streamManager.cancelStream(requestId: requestId)
        isActive = false
        
        fireEvent("cancelled", with: [:])
    }
    
    // MARK: - Event Handlers
    
    private func handleChunk(_ chunk: String) {
        guard isActive else { return }
        
        TNMLogger.Streaming.chunkReceived(
            requestId: requestId,
            size: chunk.utf8.count,
            totalSize: 0 // We don't track total in streaming
        )
        
        // Fire chunk event to JavaScript
        fireEvent("chunk", with: [
            "data": chunk
        ])
    }
    
    private func handleComplete(statusCode: Int, headers: [String: String]) {
        guard isActive else { return }
        
        isActive = false
        
        let duration = startTime.map { Date().timeIntervalSince($0) } ?? 0
        
        TNMLogger.Streaming.completed(
            requestId: requestId,
            statusCode: statusCode,
            totalBytes: 0 // Total bytes not tracked in streaming
        )
        
        // Apply response interceptors
        var modifiedStatusCode = statusCode
        var modifiedHeaders = headers
        
        interceptorManager.interceptResponse(
            statusCode: &modifiedStatusCode,
            headers: &modifiedHeaders,
            body: nil
        )
        
        // Fire complete event
        fireEvent("complete", with: [
            "statusCode": modifiedStatusCode,
            "headers": modifiedHeaders,
            "duration": duration
        ])
    }
    
    private func handleError(_ error: Error) {
        guard isActive else { return }
        
        isActive = false
        
        TNMLogger.Streaming.error(requestId: requestId, error: error)
        
        fireEvent("error", with: [
            "error": error.localizedDescription,
            "code": (error as NSError).code
        ])
    }
}

File: android/src/ti/network/manager/TiHTTPStreamProxy.java
/**
 * Ti.Network.Manager - Stream Proxy
 * JavaScript proxy for SSE streaming
 * Mirrors iOS TiHTTPStreamProxy
 */

package ti.network.manager;

import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Kroll.proxy(creatableInModule = TiNetworkManagerModule.class)
public class TiHTTPStreamProxy extends KrollProxy {

    private final TNMStreamManager streamManager;
    private final String streamId;
    private final String url;
    private final String method;
    private final Map<String, String> headers;
    private final String body;
    private final String priority;
    private boolean isActive = false;
    private int chunkCount = 0;
    private long totalBytes = 0;

    public TiHTTPStreamProxy(KrollDict params, TNMStreamManager streamManager) {
        super();

        this.streamManager = streamManager;
        this.streamId = UUID.randomUUID().toString();
        this.url = params.optString("url", "");
        this.method = params.optString("method", "GET").toUpperCase();
        this.priority = params.optString("priority", "normal");
        this.body = params.optString("body", null);

        // Headers
        if (params.containsKey("headers")) {
            @SuppressWarnings("unchecked")
            Map<String, String> headersMap = (Map<String, String>) params.get("headers");
            this.headers = headersMap;
        } else {
            this.headers = new HashMap<>();
        }

        Map<String, Object> details = new HashMap<>();
        details.put("streamId", streamId);
        details.put("url", url);
        details.put("method", method);
        TNMLogger.debug("Stream proxy created", "Stream", details);
    }

    @Kroll.method
    public void start() {
        if (isActive) {
            TNMLogger.warning("Stream already active", "Stream",
                    Map.of("streamId", streamId));
            return;
        }

        isActive = true;
        chunkCount = 0;
        totalBytes = 0;

        // Convert priority
        float urlPriority = 0.5f; // Normal
        if ("high".equals(priority)) {
            urlPriority = 0.75f;
        } else if ("low".equals(priority)) {
            urlPriority = 0.25f;
        }

        streamManager.startStream(
                streamId,
                url,
                method,
                headers,
                body,
                urlPriority,
                new TNMStreamManager.StreamCallback() {
                    @Override
                    public void onOpen() {
                        handleOpen();
                    }

                    @Override
                    public void onChunk(String data, String eventType) {
                        handleChunk(data, eventType);
                    }

                    @Override
                    public void onClose() {
                        handleClose();
                    }

                    @Override
                    public void onError(Throwable error) {
                        handleError(error);
                    }
                }
        );
    }

    @Kroll.method
    public void cancel() {
        if (!isActive) return;

        streamManager.cancelStream(streamId);
        isActive = false;

        KrollDict event = new KrollDict();
        fireEvent("cancelled", event);
    }

    private void handleOpen() {
        KrollDict event = new KrollDict();
        event.put("streamId", streamId);
        fireEvent("open", event);
    }

    private void handleChunk(String data, String eventType) {
        if (!isActive) return;

        chunkCount++;
        totalBytes += data.length();

        TNMLogger.debug("Stream chunk", "Stream", Map.of(
                "streamId", streamId,
                "chunkNumber", chunkCount,
                "size", data.length()
        ));

        KrollDict event = new KrollDict();
        event.put("streamId", streamId);
        event.put("data", data);
        event.put("eventType", eventType != null ? eventType : "message");
        event.put("chunkNumber", chunkCount);
        event.put("totalBytes", totalBytes);

        fireEvent("chunk", event);
    }

    private void handleClose() {
        isActive = false;

        TNMLogger.info("Stream completed", "Stream", Map.of(
                "streamId", streamId,
                "totalChunks", chunkCount,
                "totalBytes", totalBytes
        ));

        KrollDict event = new KrollDict();
        event.put("streamId", streamId);
        event.put("totalChunks", chunkCount);
        event.put("totalBytes", totalBytes);

        fireEvent("complete", event);
    }

    private void handleError(Throwable error) {
        if (!isActive) return;

        isActive = false;

        KrollDict event = new KrollDict();
        event.put("streamId", streamId);
        event.put("error", error.getMessage());

        fireEvent("error", event);
    }
}